swift/Workflow/Sources/SubtreeManager.swift

```diff
@@ -227,6 +240,70 @@ extension WorkflowNode.SubtreeManager {
 }
 
 
+// MARK: - Reusable Sink
+
+extension WorkflowNode.SubtreeManager {
+    fileprivate struct SinkStore {
+        private var previousSinks: [ObjectIdentifier:AnyTypedSink]
+        private (set) var usedSinks: [ObjectIdentifier:AnyTypedSink]
+
+        init(previousSinks: [ObjectIdentifier:AnyTypedSink]) {
+            self.previousSinks = previousSinks
+            self.usedSinks = [:]
+        }
+
+        mutating func findOrCreate<Action: WorkflowAction>(actionType: Action.Type) -> TypedSink<Action> {
+            let key = ObjectIdentifier(actionType)
+
+            let typedSink: TypedSink<Action>
+
+            if let previousSink = previousSinks.removeValue(forKey: key) as? TypedSink<Action> {
+                // Reused a previous sink, creating a new event pipe to send the action through.
+                previousSink.eventPipe = EventPipe()
+                typedSink = previousSink
+            } else if let usedSink = usedSinks[key] as? TypedSink<Action> {
+                // Multiple sinks using the same backing sink.
+                typedSink = usedSink
+            } else {
+                // Create a new typed sink.
+                typedSink = TypedSink<Action>()
+            }
+
+            usedSinks[key] = typedSink
+
+            return typedSink
+        }
+
+        func eventPipes() -> [EventPipe] {
+            let eventPipes = usedSinks.values.map { typedSink -> EventPipe in
+                typedSink.eventPipe
+            }
+
+            return eventPipes
+        }
+    }
+
+    fileprivate class AnyTypedSink {
+        var eventPipe: EventPipe
+
+        init() {
+            eventPipe = EventPipe()
+        }
+    }
+
+    fileprivate final class TypedSink<Action: WorkflowAction>: AnyTypedSink where Action.WorkflowType == WorkflowType {
```

Comment:
Maybe `ReusableSink` rather than `TypedSink`

Reply:
Good call on the renaming, must better description of what is does, not how it happens to be used internally.